README.md
PrimeWire (formerly LetMeWatchThis)
=========

_Keep It Simple, Stupid_ _([KISS](https://en.wikipedia.org/wiki/KISS_principle))_ _Version_

This is a plugin that creates a new channel in [Plex Media Server](https://plex.tv/) (PMS) to view content from [PrimeWire](http://www.primewire.ag).

> **Note:** the author of this plugin has no affiliation with [PrimeWire](http://www.primewire.ag) nor the owners of the content that they host.

## Features

- Watch TV & Movies
- Custom Bookmarks
- Custom Site URL
- Search for TV or Movies
- ~~Update Channel Internally~~ _(functionality disabled due to instability)_

## Install

- [Download](https://github.com/piplongrun/lmwt-kiss.bundle/archive/master.zip) and install it by following the Plex [instructions](https://support.plex.tv/hc/en-us/articles/201187656-How-do-I-manually-install-a-channel-) or the instructions below.
  - Unzip and rename the folder to "lmwt-kiss.bundle"
  - Copy "lmwt-kiss.bundle" into the PMS [Plug-ins](https://support.plex.tv/hc/en-us/articles/201106098-How-do-I-find-the-Plug-Ins-folder-) directory
  - ~~Restart PMS~~ **This is old, should not have to restart PMS.  If channel does not appear then Restart PMS**

## Issues

- Currently there are many PrieWire sites, but not all of them are structured the same.
- Current URL test is `Domain + '/watch-2741621-Brooklyn-Nine-Nine'`
  - Example: `http://www.primewire.ag/watch-2741621-Brooklyn-Nine-Nine`


Changelog.md
# ChangeLog

##### 1.1.0
- _05/10/16_
  - Fixed Vidzi Chromecast error, had to remove hls support
  - Added new valid site URL

##### 1.0.9
- _04/08/16_
  - Fixed Bookmark class error
  - Moved Unpacker code to shared code, can now be used with future service codes

##### 1.0.8
- _04/05/16_
  - Fixed Bookmark caching error
  - Improved URL Cache issue when _No Bookmarks_ pref selected

##### 1.0.7
- _04/04/16_
  - Added Vidzi.tv URL Service Code
  - Added HLS support for Vidzi.tv streams
  - Fixed typo in Prefs file

##### 1.0.6
- _03/11/16_
  - Fixed MediaVersion URL parse

##### 1.0.5
- _02/25/16_
- Added Refresh Bookmark Covers
  - Some thumbs/covers change over time, this will allow for fresh thumbs to be pulled
- Updated DumbPrefs, removed PHT
- New Message system, handles PHT and OpenPHT
- New Auth check for preferences and updater. Only plex server admin can edit prefs or update channel.
  - Can Auth against PMS or Plex.tv
  - Enable `Auth Admin through Plex.tv` if not Plex Home setup
  - Refer to [Comment_1133884](https://forums.plex.tv/discussion/comment/1133884/#Comment_1133884), [Comment_1133922](https://forums.plex.tv/discussion/comment/1133922/#Comment_1133922), and [Comment_1134571](https://forums.plex.tv/discussion/comment/1134571/#Comment_1134571) for breakdown of admin authentication

##### 1.0.4
- _01/23/16_
- Fixed Plex Web Search
- Updated DumbTools to latest
- Added Client Platform and Product checks in logs

##### 1.0.3
- _01/20/16_
- Added [DumbTools](https://github.com/coryo/DumbTools-for-Plex)
- Added Disable Bookmarks, so a wider range of URL domains can be tested.  FYI https sites don't usually work with this channel.

##### 1.0.2
- _12/30/15_
- Updated Channel Icon to match [piplongrun](https://github.com/piplongrun)

##### 1.0.1
- _12/23/15_
- Updates/New:
  - New: Moved bookmark code to Bookmark Class
  - New: Background Art
  - New: Changelog.md
  - Update: Moved to Release based [Updater](https://github.com/kolsys/plex-channel-updater)

##### 1.0.0
- _12/21/15_
- First Personal Release.
- **Note:** This is separate from [piplongrun](https://github.com/piplongrun)'s channel
- Fixes:
  - Improved Search. Should match Movie or TV Show regardless of movie/"tv show" search
- Updates/New:
  - New: Icons: Add/Remove Bookmarks, My Bookmarks, Update Channel, Search
  - New: Bookmarks
  - New: Readme.md
  - Update: Site URL (Domain) list.
  - Update: Optional custom Site URL


Contents/Code/updater.py
UPDATED_AT = 'updater.updated_at'

class GithubStrategy(object):
    def __init__(self, repo, branch = 'master'):
        super(GithubStrategy, self).__init__()
        self.repo        = repo
        self.branch      = branch
        self.archive_url = 'https://github.com/%s/archive/%s.zip'  % (repo, branch)
        self.atom_url    = 'https://github.com/%s/commits/%s.atom' % (repo, branch)

    @property
    def updated_at(self):
        ttl     = 43200 # 12 hours
        feed    = RSS.FeedFromURL(self.atom_url, cacheTime = ttl)
        updated = Datetime.ParseDate(feed.entries[0].updated)

        return updated.replace(tzinfo = None)

    def perform_update(self):
        archive = Archive.ZipFromURL(self.archive_url)
        for name in archive.Names():
            data    = archive[name]
            parts   = name.split('/')
            shifted = Core.storage.join_path(*parts[1:])
            full    = Core.storage.join_path(Core.bundle_path, shifted)

            if '/.' in name: continue

            if name.endswith('/'):
                Core.storage.ensure_dirs(full)
            else:
                Core.storage.save(full, data)

        del archive

strategy = GithubStrategy
instance = None
def init(**kwargs):
    global instance
    instance = strategy(**kwargs)

def updated_at():
    if UPDATED_AT in Dict:
        return Dict[UPDATED_AT]
    else:
        return None

def update_available():
    last_updated = updated_at()
    if last_updated is None:
        return True
    else:
        return last_updated < instance.updated_at

def PerformUpdate():
    threaded_update_if_available()
    return ObjectContainer(
        header  = L('updater.label.updating'),
        message = L('updater.response.updating')
    )

def update_if_available():
    if update_available():
        Dict[UPDATED_AT] = instance.updated_at
        Dict.Save()
        instance.perform_update()

@thread
def threaded_update_if_available():
    update_if_available()

def add_button_to(container, cb):
    if update_available():
        container.add(DirectoryObject(
            title = L('updater.label.update-now'),
            key   = Callback(cb),
            thumb = R('icon-update.png')
        ))


Contents/Code/bookmarks.py
####################################################################################################
#                                                                                                  #
#                               Plex Channel Bookmark Class                                        #
#                                                                                                  #
####################################################################################################

class Bookmark(object):

    def __init__(self, title, prefix, bm_add_icon=None, bm_rm_icon=None):
        self.title = title
        self.prefix = prefix
        self.bm_add_icon = bm_add_icon
        self.bm_rm_icon = bm_rm_icon

        Route.Connect(self.prefix + '/bookmark/add', self.add)
        Route.Connect(self.prefix + '/bookmark/remove', self.remove)
        Route.Connect(self.prefix + '/message', self.message_container)

    ################################################################################################
    def bookmark_exist(self, item_id, category):
        """Test if bookmark exist"""

        bm = Dict['Bookmarks']
        return ((True if [b['id'] for b in bm[category] if b['id'] == item_id] else False) if category in bm.keys() else False) if bm else False

    ################################################################################################
    def message_container(self, header, message):
        """Setup MessageContainer depending on Platform"""

        if Client.Platform in ['Plex Home Theater', 'OpenPHT']:
            oc = ObjectContainer(
                title1=self.title, title2=header, no_cache=True,
                no_history=True, replace_parent=True
                )
            oc.add(PopupDirectoryObject(title=header, summary=message))

            return oc
        else:
            return MessageContainer(header, message)

    ################################################################################################
    def add(self, title, url, thumb, category, item_id):
        """Add bookmark to Dict"""

        new_bookmark = {'id': item_id, 'title': title, 'url': url, 'thumb': thumb, 'category': category}
        bm = Dict['Bookmarks']

        if not bm:
            Dict['Bookmarks'] = {category: [new_bookmark]}
            Dict.Save()

            return self.message_container('Bookmarks', '\"%s\" has been added to your bookmarks.' %title)
        elif category in bm.keys():
            if (True if [b['id'] for b in bm[category] if b['id'] == item_id] else False):

                return self.message_container('Warning',
                    '\"%s\" is already in your \"%s\" bookmark list.' %(title, category))
            else:
                temp = {}
                temp.setdefault(category, bm[category]).append(new_bookmark)
                Dict['Bookmarks'][category] = temp[category]
                Dict.Save()

                return self.message_container('\"%s\" Bookmark Added' %title,
                    '\"%s\" added to your \"%s\" bookmark list.' %(title, category))
        else:
            Dict['Bookmarks'].update({category: [new_bookmark]})
            Dict.Save()

            return self.message_container('\"%s\" Bookmark Added' %title,
                '\"%s\" added to your \"%s\" bookmark list.' %(title, category))

    ################################################################################################
    def remove(self, title, item_id, category):
        """
        Remove Bookmark from Bookmark Dictionary
        If Bookmark to remove is the last Bookmark in the Dictionary,
        then Remove the Bookmark Dictionary also
        """

        bm = Dict['Bookmarks']
        if self.bookmark_exist(item_id, category):
            bm_c = bm[category]
            for i in xrange(len(bm_c)):
                if bm_c[i]['id'] == item_id:
                    bm_c.pop(i)
                    Dict.Save()
                    break

            if len(bm_c) == 0:
                del bm_c
                Dict.Save()

                return self.message_container('Remove Bookmark',
                    '\"%s\" bookmark was the last, so removed \"%s\" bookmark section' %(title, category))
            else:
                return self.message_container('Remove Bookmark',
                    '\"%s\" removed from your \"%s\" bookmark list.' %(title, category))
        elif Client.Platform in ['Plex Home Theater', 'OpenPHT']:
            return self.message_container('\"%s\" Bookmark Removed' %title,
                '\"%s\" removed from your \"%s\" bookmark list.' %(title, category))
        else:
            return self.message_container('Error',
                'ERROR: \"%s\" not found in \"%s\" bookmark list.' %(title, category))

    ################################################################################################
    def add_remove_bookmark(self, title, thumb, url, item_id, category, oc):
        """Test if bookmark exist"""

        if self.bookmark_exist(item_id, category):
            oc.add(DirectoryObject(
                key=Callback(self.remove, title=title, item_id=item_id, category=category),
                title='Remove Bookmark',
                summary='Remove \"%s\" from your Bookmarks list.' %title,
                thumb=R(self.bm_rm_icon)
                ))
        else:
            oc.add(DirectoryObject(
                key=Callback(self.add, title=title, thumb=thumb, url=url, category=category, item_id=item_id),
                title='Add Bookmark',
                summary='Add \"%s\" to your Bookmarks list.' %title,
                thumb=R(self.bm_add_icon)
                ))


Contents/Code/__init__.py
from time import sleep
from DumbTools import DumbKeyboard
from DumbTools import DumbPrefs

TITLE = 'PrimeWire'
PREFIX = '/video/lmwtkiss'

MOVIE_ICON = 'icon-movie.png'
TV_ICON = 'icon-tv.png'
BOOKMARK_ADD_ICON = 'icon-add-bookmark.png'
BOOKMARK_REMOVE_ICON = 'icon-remove-bookmark.png'

####################################################################################################
def Start():

	ObjectContainer.title1 = TITLE
	DirectoryObject.thumb = R('icon-default.png')
	HTTP.CacheTime = CACHE_1HOUR
	HTTP.Headers['User-Agent'] = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.101 Safari/537.36'

	ValidatePrefs()

####################################################################################################
@handler(PREFIX, TITLE)
def MainMenu():

	oc = ObjectContainer()

	oc.add(DirectoryObject(key=Callback(Section, title='Movies', type='movies'), title='Movies', thumb=R(MOVIE_ICON)))
	oc.add(DirectoryObject(key=Callback(Section, title='TV Shows', type='tv'), title='TV Shows', thumb=R(TV_ICON)))
	oc.add(DirectoryObject(key=Callback(BookmarksMain), title='My Bookmarks', thumb=R('icon-bookmarks.png')))
	if Client.Product in DumbPrefs.clients:
		DumbPrefs(PREFIX, oc, title='Preferences', thumb=R('icon-prefs.png'))
	else:
		oc.add(PrefsObject(title='Preferences'))
	if Client.Product in DumbKeyboard.clients:
		DumbKeyboard(PREFIX, oc, Search, dktitle='Search', dkthumb=R('icon-search.png'))
	else:
		oc.add(InputDirectoryObject(key=Callback(Search), title='Search', prompt='Search', thumb=R('icon-search.png')))

	return oc

####################################################################################################
@route(PREFIX + '/validateprefs')
def ValidatePrefs():
	"""
	Need to check urls
	if no good then block channel from running
	"""

	if (Prefs['pw_site_url'] != Dict['pw_site_url']) and not Prefs['custom_url']:
		Dict['pw_site_url'] = Prefs['pw_site_url']
	elif Prefs['custom_url']:
		Dict['pw_site_url'] = Prefs['pw_site_url_custom']
	Dict.Save()
	Log.Debug('*' * 80)

	try:
		test = HTTP.Request(Dict['pw_site_url'] + '/watch-2741621-Brooklyn-Nine-Nine', cacheTime=0).headers
		Log.Debug('* \"%s\" is a valid url' %Dict['pw_site_url'])
		Log.Debug('* \"%s\" headers = %s' %(Dict['pw_site_url'], test))
		Dict['domain_test'] = 'Pass'
	except:
		Log.Debug('* \"%s\" is not a valid domain for this channel.' %Dict['pw_site_url'])
		Log.Debug('* Please pick a different URL')
		Dict['domain_test'] = 'Fail'

	Log.Debug('*' * 80)
	Dict.Save()

####################################################################################################
def DomainTest():
	"""Setup MessageContainer if Dict[\'domain_test\'] failed"""

	if Dict['domain_test'] == 'Fail':
		return MessageContainer('Error',
			'%s is NOT a Valid Site URL for this channel.  Please pick a different Site URL.' %Dict['pw_site_url'])
	else:
		return False

####################################################################################################
@route(PREFIX + '/bookmarksmain')
def BookmarksMain():
	"""
	Setup Bookmark Main Menu.
	Seperate by TV or Movies
	"""

	bm = Dict['Bookmarks']

	if DomainTest() != False:
		return DomainTest()
	elif not bm:
		return MessageContainer('Bookmarks', 'Bookmarks list Empty')

	oc = ObjectContainer(title2='My Bookmarks', no_cache=True)

	for key in sorted(bm.keys()):
		if len(bm[key]) == 0:
			del Dict['Bookmarks'][key]
			Dict.Save()
		else:
			if 'TV' in key:
				thumb=R(TV_ICON)
			else:
				thumb=R(MOVIE_ICON)

			oc.add(DirectoryObject(
				key=Callback(BookmarksSub, category=key),
				title=key, summary='Display %s Bookmarks' %key, thumb=thumb
				))

	if len(oc) > 0:
		return oc
	else:
		return MessageContainer('Bookmarks', 'Bookmark list Empty')

####################################################################################################
@route(PREFIX + '/bookmarkssub')
def BookmarksSub(category):
	"""List Bookmarks Alphabetically"""

	bm = Dict['Bookmarks']

	if DomainTest() != False:
		return DomainTest()
	elif not category in bm.keys():
		return MessageContainer('Error',
			'%s Bookmarks list is dirty, or no %s Bookmark list exist.' %(category, category))

	oc = ObjectContainer(title2='My Bookmarks | %s' %category, no_cache=True)

	for bookmark in sorted(bm[category], key=lambda k: k['title']):
		title = bookmark['title']
		thumb = bookmark['thumb']
		url = bookmark['url']
		category = bookmark['category']
		item_id = bookmark['id']

		oc.add(DirectoryObject(
			key=Callback(MediaSubPage, title=title, category=category, thumb=thumb, item_url=url, item_id=item_id),
			title=title, thumb=thumb
			))

	if len(oc) > 0:
		oc.add(DirectoryObject(
			key=Callback(UpdateBMCovers, category=category), title='Update Bookmark Covers',
			summary='Some Cover URL\'s change over time, Use this to update covers to current URL',
			thumb=R('icon-refresh.png')
			))
		return oc
	else:
		return MessageContainer('Bookmarks', '%s Bookmarks list Empty' %category)

####################################################################################################
@route(PREFIX + '/section')
def Section(title, type='movies'):

	if DomainTest() != False:
		return DomainTest()

	if type == 'tv':
		rel_url = 'index.php?tv=&sort=%s'
	else:
		rel_url = 'index.php?sort=%s'

	oc = ObjectContainer(title2=title)

	oc.add(DirectoryObject(key=Callback(Media, title='Popular', rel_url=rel_url % ('views')), title='Popular'))
	oc.add(DirectoryObject(key=Callback(Media, title='Featured', rel_url=rel_url % ('featured')), title='Featured'))
	oc.add(DirectoryObject(key=Callback(Media, title='Highly Rated', rel_url=rel_url % ('ratings')), title='Highly Rated'))
	oc.add(DirectoryObject(key=Callback(Media, title='Recently Added', rel_url=rel_url % ('date')), title='Recently Added'))
	oc.add(DirectoryObject(key=Callback(Media, title='Latest Releases', rel_url=rel_url % ('release')), title='Latest Releases'))

	return oc

####################################################################################################
@route(PREFIX + '/media', page=int, search=bool)
def Media(title, rel_url, page=1, search=False):

	if DomainTest() != False:
		return DomainTest()

	url = Dict['pw_site_url'] + '/%s&page=%i' %(rel_url, page)

	if Dict['pw_site_url'] != Dict['pw_site_url_old']:
		Dict['pw_site_url_old'] = Dict['pw_site_url']
		Dict.Save()
		HTTP.ClearCache()

	html = HTML.ElementFromURL(url)

	oc = ObjectContainer(title2=title, no_cache=True)

	for item in html.xpath('//div[@class="index_container"]//a[contains(@href, "/watch-")]'):

		item_url = item.xpath('./@href')[0]
		item_title = item.xpath('./h2/text()')[0]
		item_thumb = item.xpath('./img/@src')[0]
		item_id = item_thumb.split('/')[-1].split('_')[0]

		if item_thumb.startswith('//'):
			item_thumb = 'http:%s' % (item_thumb)
		elif item_thumb.startswith('/'):
			item_thumb = 'http://%s%s' % (url.split('/')[2], item_thumb)

		oc.add(DirectoryObject(
			key = Callback(MediaSubPage, item_url=item_url, title=item_title, thumb=item_thumb, item_id=item_id),
			title = item_title,
			thumb = item_thumb
			))

	next_check = html.xpath('//div[@class="pagination"]/a[last()]/@href')

	if len(next_check) > 0:

		next_check = next_check[0].split('page=')[-1].split('&')[0]

		if int(next_check) > page:

			oc.add(NextPageObject(
				key = Callback(Media, title=title, rel_url=rel_url, page=page+1),
				title = 'More...'
				))

	if len(oc) > 0:
		return oc
	elif search:
		return MessageContainer('Search',
			'No Search results for \"%s\"' %title)
	else:
		return MessageContainer('Error',
			'No media for \"%s\"' %title)

####################################################################################################
@route(PREFIX + '/media/subpage')
def MediaSubPage(title, thumb, item_url, item_id, category=None):
	"""
	Split into MediaSeason (TV) or MediaVersion (Movie)
	Include Bookmark option here
	"""

	if DomainTest() != False:
		return DomainTest()

	oc = ObjectContainer(title2=title, no_cache=True)

	if not item_url.startswith('http'):
		url = Dict['pw_site_url'] + item_url
	else:
		url = item_url

	if not category:
		html = HTML.ElementFromURL(url)

		category = 'TV Shows' if html.xpath('//div[@class="tv_container"]') else 'Movies'

	if category == 'TV Shows':
		oc.add(DirectoryObject(
			key = Callback(MediaSeasons, url=url, title=title, thumb=thumb),
			title = title,
			thumb = thumb
			))
	else:
		oc.add(DirectoryObject(
			key = Callback(MediaVersions, url=url, title=title, thumb=thumb),
			title = title,
			thumb = thumb
			))

	bm = Dict['Bookmarks']

	if ((True if [b['id'] for b in bm[category] if b['id'] == item_id] else False) if category in bm.keys() else False) if bm else False:
		oc.add(DirectoryObject(
			key=Callback(RemoveBookmark, title=title, item_id=item_id, category=category),
			title='Remove Bookmark',
			summary='Remove \"%s\" from your Bookmarks list.' %title,
			thumb=R(BOOKMARK_REMOVE_ICON)
			))
	else:
		oc.add(DirectoryObject(
			key=Callback(AddBookmark, title=title, thumb=thumb, url=item_url, category=category, item_id=item_id),
			title='Add Bookmark',
			summary='Add \"%s\" to your Bookmarks list.' %title,
			thumb=R(BOOKMARK_ADD_ICON)
			))

	return oc

####################################################################################################
@route(PREFIX + '/media/seasons')
def MediaSeasons(url, title, thumb):

	if DomainTest() != False:
		return DomainTest()

	html = HTML.ElementFromURL(url)

	oc = ObjectContainer(title2=title)

	for season in html.xpath('//div[@class="tv_container"]//a[@data-id]/@data-id'):

		oc.add(DirectoryObject(
			key = Callback(MediaEpisodes, url=url, title='Season %s' % (season), thumb=thumb),
			title = 'Season %s' % (season),
			thumb = thumb
			))

	return oc

####################################################################################################
@route(PREFIX + '/media/episodes')
def MediaEpisodes(url, title, thumb):

	if DomainTest() != False:
		return DomainTest()

	html = HTML.ElementFromURL(url)

	oc = ObjectContainer(title2=title)

	for item in html.xpath('//div[@data-id="%s"]//a[contains(@href, "/tv-")]' % (title.split(' ')[-1])):

		item_title = '%s %s' % (item.xpath('.//text()')[0].strip(), item.xpath('.//text()')[1].strip().replace('â€™', "'"))

		if '0 links' in item_title.lower():
			continue

		item_url = item.xpath('./@href')[0]

		oc.add(DirectoryObject(
			key = Callback(MediaVersions, url=item_url, title=item_title, thumb=thumb),
			title = item_title,
			thumb = thumb
			))

	return oc

####################################################################################################
@route(PREFIX + '/media/versions')
def MediaVersions(url, title, thumb):

	if DomainTest() != False:
		return DomainTest()
	elif not url.startswith('http'):
		url = Dict['pw_site_url'] + url

	html = HTML.ElementFromURL(url)
	summary = html.xpath('//meta[@name="description"]/@content')[0].split(' online - ', 1)[-1].split('. Download ')[0]

	oc = ObjectContainer(title2=title)

	for ext_url in html.xpath('//a[contains(@href, "/goto.php?")]/@href'):

		url = ext_url.split('url=')[-1].split('&')[0]
		url = String.Base64Decode(url)

		if url.split('/')[2].replace('www.', '') in ['youtube.com']:
			continue

		# Trick to use the bundled Vidzi URL Service
		if 'vidzi.tv' in url:
			url = url.replace('http://', 'vidzi://')

		if URLService.ServiceIdentifierForURL(url) is not None:

			host = url.split('/')[2].replace('www.', '')

			oc.add(DirectoryObject(
				key = Callback(MediaPlayback, url=url, title=title),
				title = '%s - %s' % (host, title),
				summary = summary,
				thumb = thumb
				))

	if len(oc) < 1:
		return ObjectContainer(header='No Sources', message='No compatible sources found')
	else:
		return oc

####################################################################################################
@route(PREFIX + '/media/playback')
def MediaPlayback(url, title):

	if DomainTest() != False:
		return DomainTest()

	Log.Debug('*' * 80)
	Log.Debug('* Client.Product		= %s' %Client.Product)
	Log.Debug('* Client.Platform		= %s' %Client.Platform)
	Log.Debug('* MediaPlayback Title	= %s' %title)
	Log.Debug('* MediaPlayback URL		= %s' %url)
	Log.Debug('*' * 80)

	oc = ObjectContainer(title2=title)
	oc.add(URLService.MetadataObjectForURL(url))

	return oc

####################################################################################################
@route(PREFIX + '/media/search')
def Search(query=''):

	if DomainTest() != False:
		return DomainTest()

	oc = ObjectContainer(title2='Search for \"%s\"' %query)

	c_list = [('Movies', 'index.php?search_keywords=%s'), ('TV Shows', 'index.php?tv=&search_keywords=%s')]

	for c, url in c_list:
		rel_url = url %(String.Quote(query, usePlus=True).lower())
		if 'TV' in c:
			thumb=R(TV_ICON)
		else:
			thumb=R(MOVIE_ICON)

		oc.add(DirectoryObject(
			key=Callback(Media, title=query, rel_url=rel_url, search=True),
			title=c, thumb=thumb
			))

	return oc

####################################################################################################
@route(PREFIX + '/bookmark/add')
def AddBookmark(title, url, thumb, category, item_id):
	"""Add Bookmark"""

	if DomainTest() != False:
		return DomainTest()

	new_bookmark = {'id': item_id, 'title': title, 'url': url, 'thumb': thumb, 'category': category}
	bm = Dict['Bookmarks']

	if not bm:
		Dict['Bookmarks'] = {category: [new_bookmark]}
		Dict.Save()

		return MessageContainer('Bookmarks',
			'\"%s\" has been added to your bookmarks.' %title)
	elif category in bm.keys():
		if (True if [b['id'] for b in bm[category] if b['id'] == item_id] else False):

			return MessageContainer('Warning',
				'\"%s\" is already in your \"%s\" bookmark list.' %(title, category))
		else:
			temp = {}
			temp.setdefault(category, bm[category]).append(new_bookmark)
			Dict['Bookmarks'][category] = temp[category]
			Dict.Save()

			return MessageContainer('Bookmarks',
				'\"%s\" added to your \"%s\" bookmark list.' %(title, category))
	else:
		Dict['Bookmarks'].update({category: [new_bookmark]})
		Dict.Save()

		return MessageContainer('Bookmarks',
			'\"%s\" added to your \"%s\" bookmark list.' %(title, category))

####################################################################################################
@route(PREFIX + '/bookmark/remove')
def RemoveBookmark(title, item_id, category):
	"""
	Remove Bookmark from Bookmark Dictionary
	If Bookmark to remove is the last Bookmark in the Dictionary,
	then Remove the Bookmark Dictionary also
	"""

	bm = Dict['Bookmarks']

	if ((True if [b['id'] for b in bm[category] if b['id'] == item_id] else False) if category in bm.keys() else False) if bm else False:
		bm_c = bm[category]
		for i in xrange(len(bm_c)):
			if bm_c[i]['id'] == item_id:
				bm_c.pop(i)
				Dict.Save()
				break

		if len(bm_c) == 0:
			del bm_c
			Dict.Save()

			return MessageContainer('Remove Bookmark',
				'\"%s\" bookmark was the last, so removed \"%s\" bookmark section' %(title, category))
		else:
			return MessageContainer('Remove Bookmark',
				'\"%s\" removed from your \"%s\" bookmark list.' %(title, category))

####################################################################################################
@route(PREFIX + '/bookmark/update/covers')
def UpdateBMCovers(category):
	"""
	Some Cover URL\'s change over time
	Use this to update covers to current URL
	"""

	bm = Dict['Bookmarks']
	bookmark_list = []
	for bookmark in sorted(bm[category], key=lambda k: k['title']):
		title = bookmark['title']
		thumb = bookmark['thumb']
		url = bookmark['url']
		category = bookmark['category']
		item_id = bookmark['id']

		bookmark_list.append(
			{'id': item_id, 'title': title, 'url': url, 'thumb': thumb, 'category': category}
			)

	Thread.Create(update_bm_thumb, bookmark_list=bookmark_list)

	return MessageContainer('Update Bookmark Covers',
		'\"%s\" Bookmark covers will be updated' %category)

####################################################################################################
def update_bm_thumb(bookmark_list=list):
	"""
	Pull Fresh Cover URL
	Update Bookmark by deleting first, then re-add Bookmark with new values
	"""

	for nbm in bookmark_list:
		category = nbm['category']
		item_id = nbm['id']
		item_url = nbm['url']

		if not item_url.startswith('http'):
			url = Dict['pw_site_url'] + item_url
		else:
			url = item_url

		html = HTML.ElementFromURL(url)
		Log.Debug('*' * 80)
		Log.Debug('* Updating \"%s\" Bookmark Cover' %nbm['title'])
		thumb = html.xpath('//meta[@property="og:image"]/@content')[0]
		if not thumb.startswith('http'):
			thumb = 'http:' + thumb

		Log.Debug('* thumb = %s' %thumb)
		nbm.update({'thumb': thumb})

		# delete bm first so we can re-append it with new values
		bm_c = Dict['Bookmarks'][category]
		for i in xrange(len(bm_c)):
			if bm_c[i]['id'] == item_id:
				bm_c.pop(i)
				Dict.Save()
				break

		# now append updatd bookmark to correct category
		temp = {}
		temp.setdefault(category, Dict['Bookmarks'][category]).append(nbm)
		Dict['Bookmarks'][category] = temp[category]
		Dict.Save()

		timer = int(Util.RandomInt(2,5) + Util.Random())
		sleep(timer)  # sleep (0-30) seconds inbetween cover updates

	return
